### PR TITLE
ci: github actions for testing and linting

### PR DESCRIPTION
**Pull request by gary-lgy**
Example CI run for push to main: https://github.com/gary-lgy/kubeadmiral/actions/runs/4718894465
Example CI run for PR: https://github.com/gary-lgy/kubeadmiral/actions/runs/4718489828

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+#           only-new-issues: true
+
+# This workflow will build a golang project
+# For more information see: https://docs.github.com/en/actions/automating-builds-and-tests/building-and-testing-go
+
+name: Go
+
+on:
+  push:
+    branches: ["main"]
```
Comment (SOF3) [inline .github/workflows/ci.yml:64]: Why do we limit branches to `main`?

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+      - name: Set up Go
+        uses: actions/setup-go@v3
+        with:
+          go-version: 1.19
+
+      - name: Build
+        run: go build -v ./...
+
+      - name: Test
+        run: go test -v ./...
```
Comment (SOF3) [inline .github/workflows/ci.yml:83]: Should we `-race` here?

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    steps:
+      - uses: actions/checkout@v3
+
+      - name: Set up Go
+        uses: actions/setup-go@v3
+        with:
+          go-version: 1.19
+
+      - name: Build
+        run: go build -v ./...
```
Comment (SOF3) [inline .github/workflows/ci.yml:80]: Is building necessary here? Wouldn't `go test` already reveal possible compile errors?

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -0,0 +1,83 @@
+# name: CI
```
Comment (SOF3) [inline .github/workflows/ci.yml:1]: Why two workflows here?

**Review by SOF3 — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -0,0 +1,51 @@
+name: CI
+
+on:
+  push:
+  pull_request:
+
+# TODO: do we need this?
+# permissions:
+#   contents: read # for actions/checkout to fetch code
```
Comment (SOF3) [inline .github/workflows/ci.yml:9]: afaik this is pretty much implied. most workflows checkout without specifying this anyway.

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    steps:
+      - uses: actions/checkout@v3
+
+      - name: Setup Go ${{ matrix.go-version }}
+        uses: actions/setup-go@v4
+        with:
+          go-version: ${{ matrix.go-version }}
+
+      - name: Build
+        run: make manager
```
Comment (SOF3) [inline .github/workflows/ci.yml:31]: Why do we need to build manager here?

Comment (gary-lgy) [reply .github/workflows/ci.yml:31]: `go test` only builds packages that have unit tests, iirc.

Comment (gary-lgy) [reply .github/workflows/ci.yml:31]: Looks like I'm wrong. From `go help test`:
> 'Go test' recompiles each package along with any files with names matching
the file pattern "*_test.go".

File: .github/workflows/ci.yml (hunk 1 above)
Comment (gary-lgy) [reply .github/workflows/ci.yml]: This is to make sure CI doesn't run twice for each push to an open PR

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    steps:
+      - uses: actions/checkout@v3
+
+      - uses: actions/setup-go@v4
+        with:
+          go-version: "1.19"
+
+      - run: |
+          make generate -B
+          git diff --exit-code && exit 0
```
Comment (SOF3) [inline .github/workflows/ci.yml:64]: If I remember correctly, github runs commands with `-e` enabled, so this line would just fail directly if there is diff. But it should be obvious enough anyway.

**Review by SOF3 — APPROVED** [review]

**Pull request merged**